### PR TITLE
Fiji-MaMuT tracking export fix

### PR DESCRIPTION
**Pull request by wolny**
Loading the results of the tracking results exported with the Fiji-MaMuT plugin in Fiji was broken (see https://github.com/ilastik/ilastik/issues/1755 for more details).

This PR fixes the issue by relying on the user to provide the BigDataViewer XML file during the export from ilastik. This way we don't have to replicate the BigDataViewer functionality neither in ilastik nor in ilastik4ij. Also as a side effect the user doesn't need to have ilastik4ij plugin installed in Fiji in order to open the exported MaMuT file.
BigDataViewer XML file can be specified in headless via `--big_data_viewer_xml_file` option.

TODO:
- [x] Fix tests

fijiexport1
fijiexport2
fijiexport3
fijiexport4

### CODE REVIEW

File: ilastik/plugins_default/tracking_mamut_export.py [hunk 1]
```diff
@@ -1,10 +1,10 @@
-from builtins import range
 import os.path
+from builtins import range
```
[the remark this answers is not in the document]
Comment (wolny) [reply ilastik/plugins_default/tracking_mamut_export.py]: nice catch! done

File: ilastik/plugins_default/tracking_mamut_export.py [hunk 2]
```diff
@@ -39,32 +39,19 @@ def checkFilesExist(self, filename):
...
-        """Export the tracking solution stored in the hypotheses graph to three files: 
-        Firstly, a _raw.h5 file containing the data, and a _bdv.xml file
-        that allows the raw data to be be displayed in Fiji's BigDataViewer.
-        Secondly, it exports an _mamut.xml file that contains the tracks for visualization and proof-reading in MaMuT. 
+    def export(self, filename, hypothesesGraph, objectFeaturesSlot, bdvFilepathSlot, **kwargs):
+        """Export the tracking solution stored in the hypotheses graph to MaMuT XML file.
+        Creates an _mamut.xml file that contains the tracks for visualization and proof-reading in MaMuT.
 
-        :param filename: string of the FILE where to save the result (different .xml files and a .h5 file with suffixes)
+        :param filename: string of the FILE where to save the result (different .xml files were)
```
[the remark this answers is not in the document]
Comment (wolny) [reply ilastik/plugins_default/tracking_mamut_export.py]: LOL :see_no_evil:

File: ilastik/plugins_default/tracking_mamut_export.yapsy-plugin [hunk 3]
```diff
@@ -6,4 +6,4 @@ Module = tracking_mamut_export
...
+Description = Plugin to export the ilastik tracking results to Fiji's MaMuT <br> <br> <b>Usage: </b> Select the folder where the MaMuT XML file will be saved and provide the path to the BigDataViewer XML file (if you don't have one already export your dataset in Fiji first with <i>Plugins->BigDataViewer->Export Current Image as XML/HDF5</i>): <ul> <li> <b>MaMut Export </b> to be opened with FiJi Mamut Plugin in order to visualise the tracking results. Has the <i> _mamut.xml </i> extension </ul> <br><br> To open the saved result in Fiji, click <i>Plugins->MaMuT->Open MaMuT annotation</i> and select the exported file with the <i> _mamut.xml </i> suffix.
```
[the remark this answers is not in the document]
Comment (wolny) [reply ilastik/plugins_default/tracking_mamut_export.yapsy-plugin]: fair enough

File: ilastik/workflows/tracking/conservation/conservationTrackingWorkflow.py [hunk 4]
```diff
@@ -391,7 +392,10 @@ def post_process_lane_export(self, lane_index, checkOverwriteFiles=False):
                     return True
 
                 self.dataExportApplet.progressSignal(-1)
-                exportStatus = self.trackingApplet.topLevelOperator.getLane(lane_index).exportPlugin(filename, exportPlugin, checkOverwriteFiles)
+                exportStatus = self.trackingApplet.topLevelOperator\
+                    .getLane(lane_index)\
```
[the remark this answers is not in the document]
Comment (wolny) [reply ilastik/workflows/tracking/conservation/conservationTrackingWorkflow.py]: I've limit the length of the line in my IDE to 79 (see: https://www.python.org/dev/peps/pep-0008/#maximum-line-length). I try to adhere to this rule in every new line of code. The one you've recommended would be more than 79...

[the remark this answers is not in the document]
Comment (wolny) [reply ilastik/workflows/tracking/conservation/conservationTrackingWorkflow.py]: Though I don't mind changing the line if that would be more in line with the conventions we're using in our codebase

[the remark this answers is not in the document]
Comment (k-dominik) [reply ilastik/workflows/tracking/conservation/conservationTrackingWorkflow.py]: Oh, sorry. I think before your time we informally agreed on a 120 character line length...

File: ilastik/plugins.py [hunk 5]
```diff
@@ -161,15 +161,18 @@ def checkFilesExist(self, filename):
         ''' Check whether the files we want to export (when appending the base filename) are already present '''
         return False
 
-    def export(self, filename, hypothesesGraph, objectFeaturesSlot, labelImageSlot, rawImageSlot):
+    def export(self, filename, hypothesesGraph, objectFeaturesSlot, labelImageSlot, rawImageSlot, bdvFilepathSlot):
```
[the remark this answers is not in the document]
Comment (wolny) [reply ilastik/plugins.py]: That's a really cool trick. Thanks for pointing this out, I wasn't aware that you could enforce keyword only args with it.

[the remark this answers is not in the document]
Comment (wolny) [reply ilastik/plugins.py]: done, please have another look if that's what you had in mind :white_check_mark:

File: ilastik/plugins_default/tracking_mamut_export.yapsy-plugin (hunk 3 above)
[the remark this answers is not in the document]
Comment (wolny) [reply ilastik/plugins_default/tracking_mamut_export.yapsy-plugin]: :white_check_mark:

File: ilastik/workflows/tracking/conservation/conservationTrackingWorkflow.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wolny) [reply ilastik/workflows/tracking/conservation/conservationTrackingWorkflow.py]: :white_check_mark:

File: ilastik/plugins_default/tracking_mamut_export.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wolny) [reply ilastik/plugins_default/tracking_mamut_export.py]: :white_check_mark:

**Comment by wolny** [comment]
Thanks for poking the tests and the review.

**Comment by chaubold** [comment]
Hey guys! 

> This PR fixes the issue by relying on the user to provide the BigDataViewer XML file during the export from ilastik. This way we don't have to replicate the BigDataViewer functionality neither in ilastik nor in ilastik4ij. Also as a side effect the user doesn't need to have ilastik4ij plugin installed in Fiji in order to open the exported MaMuT file.

Just my 2 cents about the reasoning why it was initially done the other way round: we wanted to prevent requiring a copy of the image data once for BDV and once for ilastik, which becomes really painful if your data is huge. Hence we taught BDV how to read ilastik's (up to) 5D HDF5 volumes directly...

But, I'm happy to see this is working at all ;-)

**Pull request merged**